src/tests/testCacheManager.cc

```diff
@@ -66,3 +75,150 @@ testCacheManager::testRegister()
+        "? foo=bar",
+        "? &",
+        "?& ",
+        "?=&",
+        "?&=",
+        "? &&&",
+        "?& &&",
+        "?&& &",
+        "?=&&&",
+        "?&=&&",
+        "?&&=&"
+    };
+
+    const std::vector<const char *> validFragments = {
+        "",
+        "#",
+        "##",
+        "#?a=b",
+        "#fragment"
+    };
+
+    unsigned success = 0;
+    for (const auto &scheme : validSchemes) {
+        mgrUrl.setScheme(scheme);
+
+        for (const auto *magic : magicPrefixes) {
+
+            // all schemes except cache_object require magic path prefix bytes
+            if (scheme != AnyP::PROTO_CACHE_OBJECT && strlen(magic) <= 2)
+                continue;
+
+            /* Check the parser accepts all the valid cases */
+
+            for (const auto *action : validActions) {
+                for (const auto *param : validParams) {
+                    for (const auto *frag : validFragments) {
+                        try {
+                            SBuf bits;
+                            bits.append(magic);
+                            bits.append(action);
+                            bits.append(param);
+                            bits.append(frag);
+                            mgrUrl.path(bits);
+
+                            (void)mgr->ParseUrl(mgrUrl);
+                            CPPUNIT_ASSERT(++success);
+                        } catch (...) {
+                            std::cerr << std::endl
+                                      << "FAIL: " << mgrUrl
+                                      << Debug::Extra << "error: " << CurrentException << std::endl;
+                            CPPUNIT_FAIL("bad result");
+                        }
+                    }
+                }
+            }
+
+            /* Check that invalid parameters are rejected */
+
+            for (const auto *action : validActions) {
+                for (const auto *param : invalidParams) {
+                    for (const auto *frag : validFragments) {
+                        static const auto err = "accepted malformed input";
+                        try {
+                            SBuf bits;
+                            bits.append(magic);
+                            bits.append(action);
+                            bits.append(param);
+                            bits.append(frag);
+                            mgrUrl.path(bits);
+
+                            (void)mgr->ParseUrl(mgrUrl);
+
+                            std::cerr << std::endl
+                                      << "FAIL: " << mgrUrl
+                                      << Debug::Extra << "error: should be rejected due to '" << param << "'" << std::endl;
+                            CPPUNIT_FAIL(err);
+                        } catch (const TextException &e) {
+                            CPPUNIT_ASSERT(++success);
+                            if (strcmp(e.what(), err) == 0)
+                                throw;
+                        }
```

Comment:
I do not understand the logic here because it is not clear where `accepted malformed input` (i.e. `err`) value might come from. What are we trying to detect/test by using `err`? If these are some manual testing leftovers, then we can remove this to simplify:

```
                        } catch (...) {
                            ... report CurrentException here, just in case it is not really what we expect ...
                            CPPUNIT_ASSERT(true); // or perhaps there is CPPUNIT_SUCCESS()?
                        }
```

The above sketch also prints the caught exception. It is never a good idea to hide exceptions, but I am not sure whether CppUnit supports some kind of debugging mode where they can be reported without polluting regular output too much. On the other hand, the current "make check" output is already so polluted that perhaps I should not be worried about additional pollution at this point.

Also, we should not be counting successes if we are not using them anywhere. KISS.

Reply:
CppUnit uses exceptions. Its "ASSERT" macros and such actually are just throwing. Only exceptions which reach the test harness are handled as errors in the code.

So ..
 a) when our code properly throws an exception for the invalid input we need to swallow it to prevent cppunit halting the build. That is what the explicit try-catch is for.

 b) when our code wrongly accepts an invalid input we need to trigger a cppunit exception. That is CPPUNIT_FAIL().
 However our try-catch also swallows these. So we need to identify and re-throw. Use of a static err string allows that efficiently without involving memory management risks.

Reply:
Just to avoid misunderstanding: This change request remains unaddressed. Mys last https://github.com/squid-cache/squid/pull/788#discussion_r608733550 enumerates two simple specific solutions that I see.

Reply:
Please note the code as at a24c52a passes semaphore and jenkins builds claiming SUCCESS.

The a24c52a code contains an intentional valid URL in the set intended to prove parser throws on bad cases. On this false-positive the parser will return normally and CPPUNIT_FAIL() throw a char* type of exception. This demonstrates that "catch (const TextException &)" is not sufficient to let cppunit thrown exceptions actually reach the cppunit handlers.

Reply:
> "catch (const TextException &)" is not sufficient to let cppunit thrown exceptions actually reach the cppunit handlers.

I cannot explain that alleged outcome without research. The above conclusion matches neither my C++ understanding nor my quick [test](https://cpp.godbolt.org/z/fhz63orfr), but there are several unknowns here that may lead to misleading conclusions. Even though this mystery is not related to any of the two solutions I have [proposed](https://github.com/squid-cache/squid/pull/788#discussion_r608733550) on this thread, I will study it to ensure I am not missing something for future work.

Reply:
Your second proposal has worked AFAICT. So going with that.